src/adapter/src/coord/introspection.rs

```diff
@@ -32,6 +32,39 @@ use crate::notice::AdapterNotice;
 use crate::session::Session;
 use crate::{rbac, AdapterError};
 
+fn constant_or_unmaterializable_scalar(scalar: &MirScalarExpr) -> bool {
+    match scalar {
+        mz_expr::MirScalarExpr::Column(_)
+        | mz_expr::MirScalarExpr::Literal(_, _)
+        | mz_expr::MirScalarExpr::CallUnmaterializable(_) => true,
+        mz_expr::MirScalarExpr::CallUnary { .. }
+        | mz_expr::MirScalarExpr::CallBinary { .. }
+        | mz_expr::MirScalarExpr::CallVariadic { .. }
+        | mz_expr::MirScalarExpr::If { .. } => false,
+    }
+}
+
+fn constant_or_unmaterializable(expr: &MirRelationExpr) -> bool {
+    match expr {
+        MirRelationExpr::ArrangeBy { input, keys } => {
+            constant_or_unmaterializable(input)
+                && keys
+                    .iter()
+                    .all(|keys| keys.iter().all(constant_or_unmaterializable_scalar))
+        }
+        MirRelationExpr::Constant { .. } | MirRelationExpr::Get { .. } => true,
+        MirRelationExpr::Map { input, scalars } => {
+            constant_or_unmaterializable(input)
+                && scalars.iter().all(constant_or_unmaterializable_scalar)
+        }
+        MirRelationExpr::Let { value, body, .. } => {
```

Comment:
Are `Let`s constant? I don't know exactly what they do but the docs say they make a temp dataflow which sounds non-constanty.

Reply:
`select 1` turns into `Select(SelectPlan { source: Let { id: LocalId(0), value: Constant { rows: Ok([(Row{[]}, 1)]), typ: RelationType { column_types: [], keys: [] } }, body: Project { input: Map { input: Get { id: Local(LocalId(0)), typ: RelationType { column_types: [], keys: [[]] } }, scalars: [Literal(Ok(Row{[Int32(1)]}), ColumnType { scalar_type: Int32, nullable: false })] }, outputs: [0] } }, when: Immediately, finishing: RowSetFinishing { order_by: [], limit: None, offset: 0, project: [0] }, copy_to: None })` which is why i included `Let`. My thinking is that if the value and body are constant nothing interesting can sneak in.

Reply:
Oh that also has a `Get` in it from the local `Let`. I thiiink the optimizer does the squashing there to get rid of both, but we're trying to figure out how to never even call the optimizer, right? Tricky.

Reply:
Wait, is it ok to call the optimizer and use an `OptimizedMirRelationExpr` instead?